ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/sse/DefaultClientSSESession.kt

```diff
@@ -4,36 +4,121 @@
 @OptIn(InternalAPI::class)
+@Deprecated("It should be marked with `@InternalAPI`, please use `ClientSSESession` instead")
 public class DefaultClientSSESession(
     content: SSEClientContent,
     private var input: ByteReadChannel,
-    override val coroutineContext: CoroutineContext,
+    override val coroutineContext: CoroutineContext
 ) : SSESession {
     private var lastEventId: String? = null
     private var reconnectionTimeMillis = content.reconnectionTime.inWholeMilliseconds
     private val showCommentEvents = content.showCommentEvents
     private val showRetryEvents = content.showRetryEvents
+    private var needToReconnect = content.allowReconnection
+    private val maxRetries = content.maxRetries
 
-    private val _incoming = channelFlow {
-        while (true) {
-            val event = input.parseEvent() ?: break
+    private val initialRequest = content.initialRequest
+
+    private val clientForReconnection = initialRequest.attributes[SSEClientForReconnectionAttr]
+
+    public constructor(
+        content: SSEClientContent,
+        input: ByteReadChannel
+    ) : this(content, input, content.callContext + Job() + CoroutineName("DefaultClientSSESession"))
+
+    private var _incoming = flow {
+        while (this@DefaultClientSSESession.coroutineContext.isActive) {
+            while (this@DefaultClientSSESession.coroutineContext.isActive) {
+                val event = input.parseEvent() ?: break
 
-            if (event.isCommentsEvent() && !showCommentEvents) continue
-            if (event.isRetryEvent() && !showRetryEvents) continue
+                if (event.isCommentsEvent() && !showCommentEvents) continue
+                if (event.isRetryEvent() && !showRetryEvents) continue
 
-            send(event)
+                emit(event)
+            }
+
+            if (needToReconnect) {
+                doReconnection()
+            } else {
+                close()
+            }
+        }
+    }
+
+    init {
+        coroutineContext.job.invokeOnCompletion {
+            close()
+        }
+    }
+
+    private suspend fun doReconnection() {
+        var retries = 1
+        while (retries <= maxRetries) {
+            try {
+                withContext(coroutineContext) {
+                    input.cancel()
+
+                    delay(reconnectionTimeMillis)
+
+                    val reconnectionRequest = getRequestForReconnection()
+                    LOGGER.trace(
+                        "Sending SSE request ${reconnectionRequest.url} (attempt ${retries + 1}/${maxRetries + 1})"
+                    )
+
+                    val reconnectionResponse = clientForReconnection.execute(reconnectionRequest).response
+                    LOGGER.trace("Receive response for reconnection SSE request to ${reconnectionRequest.url}")
+                    checkResponse(reconnectionResponse)
+
+                    if (reconnectionResponse.status == HttpStatusCode.NoContent) {
+                        needToReconnect = false
```

Comment:
Have I understood it correctly that after we receive status code 204 once, we stop any further reconnections? I'm not sure I understand the logic behind this

Reply:
Yes, in the [documentation](https://html.spec.whatwg.org/multipage/server-sent-events.html) written the next:
> Clients will reconnect if the connection is closed; a client can be told to stop reconnecting using the HTTP 204 No Content response code.